Don't broadcast old messages to the feed

Tristan-Wilson opened the pull request:
Fixes NIT-3702

During node synchronization and catch-up, broadcasting all historical messages can flood connected feed clients with redundant data. This change introduces a check before broadcasting that only messages that are newer than the network's head minus the last two batches for backlog maintenance.

The new behavior is to only broadcast messages within the last 2 batches when not synced. New sequencer messages are always broadcast regardless of sync state (this is the sequencer feed, so messages produced by the sequencer must be in it).

This feature is enabled by default so operators don't need to make any config changes to start using it. It has hot-reloadable config flag that can be used to disable it (ie still broadcast historical messages during sync):
```
--node.transaction-streamer.broadcast-during-sync
```

Summary of new broadcasting behavior:
- New sequencer messages: always broadcast
- When synced: broadcast everything normally
- When syncing: only broadcast last 2 batches
- Fail-safe: any error condition defaults to broadcasting

Tristan-Wilson committed:
Don't broadcast old messages to the feed

During node synchronization and catch-up, broadcasting all historical messages
can flood connected feed clients with redundant data. This change introduces
a check before broadcasting that only messages that are newer than the
network's head minus the last two batches for backlog maintenance.

The new behavior is to only broadcast messages within the last 2 batches
when not synced. New sequencer messages are always broadcast regardless
of sync state (this is the sequencer feed, so messages produced by the
sequencer must be in it).

This feature is enabled by default so operators don't need to make any
config changes to start using it. It has hot-reloadable config flag that
can be used to disable it:
```
--node.transaction-streamer.disable-broadcast-during-sync
```

Summary of new broadcasting behavior:
- New sequencer messages: always broadcast
- When synced: broadcast everything normally
- When syncing: only broadcast last 2 batches
- Fail-safe: any error condition defaults to broadcasting

eljobe committed:
Merge branch 'master' into fix-historical-feed-flood

diegoximenes commented on arbnode/node.go:
```diff
@@ -805,11 +805,12 @@ func getTransactionStreamer(
 ) (*TransactionStreamer, error) {
 	transactionStreamerConfigFetcher := func() *TransactionStreamerConfig { return &configFetcher.Get().TransactionStreamer }
-	txStreamer, err := NewTransactionStreamer(ctx, arbDb, l2Config, exec, broadcastServer, fatalErrChan, transactionStreamerConfigFetcher, &configFetcher.Get().SnapSyncTest)
+	txStreamer, err := NewTransactionStreamer(ctx, arbDb, l2Config, exec, broadcastServer, syncMonitor, fatalErrChan, transactionStreamerConfigFetcher, &configFetcher.Get().SnapSyncTest)
```
SyncMonitor depends on TransactionStreamer today, and this creates a circular dependency between them 😬.

This indicates to me that they could be better placed in a single component.
How about moving SyncMonitor logic to TransactionStreamer and then dropping SyncMonitor?

I created this [linear task](https://linear.app/offchain-labs/issue/NIT-3649/instead-of-executions-syncmonitorfullsyncprogressmap-and in order) in order to move logic from SyncMonitor to ConsensusExecutionSyncer, but it is OK to move it TransactionStreamer 🙂.
ConsensusExecutionSyncer already depends on TransactionStreamer, and can call it for what it needs.

Tristan-Wilson replied:
I made a PR to remove the circular dependency: https://github.com/OffchainLabs/nitro/pull/3538. If you like the approach then I'll rebase this PR ono that.

diegoximenes commented on arbnode/transaction_streamer.go:
```diff
@@ -205,6 +213,10 @@ func (s *TransactionStreamer) SetInboxReaders(inboxReader *InboxReader, delayedB
 	s.delayedBridge = delayedBridge
+
+	if s.syncMonitor != nil && inboxReader != nil && inboxReader.tracker != nil {
+		s.broadcastChecker = NewBroadcastSyncChecker(s.syncMonitor, inboxReader.tracker)
```
This is also introducing a circular dependency:
TransactionStreamer -> BroadcastSyncChecker -> SyncMonitor -> TransactionStreamer

This also indicates to me that all of them are better placed in a single component.
How about moving BroadcastSyncChecker logic to TransactionStreamer too?

diegoximenes commented on arbnode/transaction_streamer.go:
```diff
@@ -205,6 +213,10 @@ func (s *TransactionStreamer) SetInboxReaders(inboxReader *InboxReader, delayedB
 	s.inboxReader = inboxReader
 	s.delayedBridge = delayedBridge
+
+	if s.syncMonitor != nil && inboxReader != nil && inboxReader.tracker != nil {
```
s.inboxReader and s.inboxReader.tracker nil checks here are redundant, they will never be nil here, right?

Tristan-Wilson replied:
Removed the separate BroadcastSyncChecker.

diegoximenes commented on arbnode/broadcast_sync_checker.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+// ShouldBroadcast determines if messages should be broadcast based on sync state and message position
+func (b *BroadcastSyncChecker) ShouldBroadcast(firstMsgIdx arbutil.MessageIndex, msgCount int) bool {
```
How about adding more information, through comments, of what motivated ShouldBroadcast's strategy, e.g., writing a comment here based on the PR description?

Tristan-Wilson replied:
Added more comments.

diegoximenes commented on arbnode/broadcast_sync_checker_test.go:
```diff
@@ -0,0 +1,211 @@
+		expected          bool
+		description       string
+	}{
+		{
+			name:        "empty batch - never broadcast",
+			msgCount:    0,
+			expected:    false,
+			description: "Empty message batches should never be broadcast",
+		},
```
ShouldBroadcast could be returning true due to batchCount being equal to zero too.
How about using the same parameters used by "not synced - last message at threshold - broadcast" case, but setting batchInfoError to true instead?

diegoximenes commented on arbnode/broadcast_sync_checker_test.go:
```diff
@@ -0,0 +1,211 @@
+			expected:    true,
+			description: "When node is synced, should always broadcast",
+		},
+		{
+			name:           "batch info error - always broadcast",
+			synced:         false,
+			msgCount:       5,
+			batchInfoError: true,
+			expected:       true,
+			description:    "When batch info provider has errors, fail open and broadcast",
+		},
```
Same comment that I added to the "empty batch - never broadcast" case applies here.

diegoximenes commented on arbnode/broadcast_sync_checker_test.go:
```diff
@@ -0,0 +1,211 @@
+	}
+}
+
+func TestBroadcastSyncCheckerEdgeCases(t *testing.T) {
```
nitpick: why not defining those threshold tests in TestBroadcastSyncChecker test cases?

Tristan-Wilson replied:
Moved them into the main test

diegoximenes reviewed (changes requested):
Nice 🙂

Tristan-Wilson committed:
Merge branch 'master' into fix-historical-feed-flood

Tristan-Wilson commented:
Sorry for the assignee spam, was looking at the wrong PR

Tristan-Wilson committed:
Merge branch 'master' into fix-historical-feed-flood

Tristan-Wilson commented:
Changed base to https://github.com/OffchainLabs/nitro/pull/3538 which removes the circular dependency, will work on fixing conflicts and addressing the comments.

Tristan-Wilson committed:
Remove TransactionStreamer<->SyncMonitor circ dep

TransactionStreamer now pushes updates to the SyncMonitor of the
currentMessageCount and feedPendingMessageCount every time it updates
either of these in its own state. This means SyncMonitor doesn't need to
have a reference to TransactionStreamer.

The logic for checking if the TransactionStreamer should broadcast to
the feed has been moved into TransactionStreamer, with the core decision
logic split out for testing.

Tristan-Wilson committed:
Merge branch 'master' into fix-historical-feed-flood

Tristan-Wilson committed:
Merge branch 'master' into fix-historical-feed-flood

Tristan-Wilson committed:
Increase timeout for NodeB for path scheme CI failures

Tristan-Wilson committed:
Allow broadcast during sync for path tests

tsahee committed:
Merge branch 'master' into fix-historical-feed-flood

tsahee commented on arbnode/node.go:
```diff
@@ -210,6 +211,10 @@ func ConfigDefaultL1NonSequencerTest() *Config {
 	config.BlockValidator.ValidationServerConfigs = []rpcclient.ClientConfig{{URL: ""}}
 	config.Bold.MinimumGapToParentAssertion = 0
+	// Disable broadcast during sync only for path tests as it causes test timeouts
+	if env.GetTestStateScheme() == rawdb.PathScheme {
```
Why does that happen?
Also it is possible for a test to use pathdb even if env.GetTestStateScheme() != rawdb.PathScheme.
see builder.RequireScheme

Tristan-Wilson committed:
Merge remote-tracking branch 'origin/master' into fix-historical-feed-flood

Resolved transaction_streamer.go: kept both config options and adapted
to new broadcaster API (BroadcastFeedMessages).

pmikolajczyk41 committed:
Merge branch 'master' into fix-historical-feed-flood

Tristan-Wilson committed:
Merge remote-tracking branch 'origin/master' into fix-historical-feed-flood

Tristan-Wilson committed:
Enable historical feed flood protection for tests

This better simulates production conditions, even if it currently causes
more tests to fail.

Tristan-Wilson committed:
Fix deadlock in TransactionStreamer when updating SyncMonitor

AddBroadcastMessages, ReorgAtAndEndBatch, and writeMessages all hold
insertionMutex when calling updateSyncMonitor(). However, updateSyncMonitor()
called FeedPendingMessageCount() which attempts to acquire insertionMutex
again. Since Go mutexes are not reentrant, this caused a deadlock.

The fix introduces updateSyncMonitorLocked() for use when the mutex is
already held, which calls feedPendingMessageCountImpl() instead of the
locking FeedPendingMessageCount().

Tristan-Wilson committed:
Remove negative from broadcast-during-sync cli opt

Having "disable" in binary command line flags is confusing, so the
disable-broadcast-during-sync flag was renamed to broadcast-during-sync
and the logic was inverted.

Also added changelog.

diegoximenes commented on arbnode/transaction_streamer.go:
```diff
@@ -1205,14 +1241,82 @@ func (s *TransactionStreamer) writeMessage(msgIdx arbutil.MessageIndex, msg arbo
+
+	// We're not synced, so check if these messages are within the 2 batch backlog threshold.
+	// If we can't determine the threshold (no metadata or not enough batches), fail open and broadcast.
+	if !haveBatchMetadata || batchCount < 2 {
```
In which cases haveBatchMetadata can be false and this not be related to a DB error?

The fact that this variable exists got me a little bit confused.
Not sure why the policy is something like always, or never, broadcast if this procedure failed to retrieve some information that should exist, such as batchThresholdMsgCount

diegoximenes commented on arbnode/transaction_streamer_test.go:
```diff
@@ -0,0 +1,251 @@
+			batchThresholdMsgCount: 100,
+			haveBatchMetadata:      true,
+			expectedBroadcast:      false,
+			description:            "Single message just before threshold should not broadcast",
```
sanity check: shouldn't this be also broadcasted?
batchThresholdMsgCount being 100 means that batchThreshold has messages up until msgIdx 99, right?
If that is true then it should be broadcasted, since lastMsgIdx is 99, which belongs to the batchThreshold.
